gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/datafetcher/QueryServiceEntityFetcher.java

```diff
@@ -273,9 +273,21 @@ public EntityFetcherResponse getEntitiesAndAggregatedMetrics(
     QueryRequest.Builder builder =
         constructSelectionQuery(requestContext, entitiesRequest, entityIdAttributes, aggregates);
 
-    // Order by, limit and offset from the request.
-    builder.setLimit(entitiesRequest.getLimit());
-    builder.setOffset(entitiesRequest.getOffset());
+    // If there is more than one groupBy column, we cannot set the same limit that came
+    // in the request since that might return less entities than needed when the same
+    // entity has different values for the other group by columns. Example: A service entity's
+    // name changes and that will now have two different names.
+    // For now, we pass a high value of limit in this case so that we get all the entities.
+    // Limit has to be applied post the query in this case. Setting offset also might be wrong
+    // here, hence not setting it.
+    if (builder.getGroupByCount() > 1) {
+      builder.setLimit(QueryServiceClient.DEFAULT_QUERY_SERVICE_GROUP_BY_LIMIT);
```

Comment:
Can you check the ExecutionVisitor and see if you need to do the limit and offset yourself?

Reply:
Yeah I've checked that already. There, we just copy the limit that comes in the request. This is more of OLAP specific implementation detail, needn't spread outside QSEFetcher class.